File: partition_list.py
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of linked list.
    @param x: an integer
    @return: a ListNode 
    """
    def partition(self, head, x):
        if not head:
            return None
            
        # to keep tracking smaller and greater sublists, declare two dummy node
        small_dummy = ListNode(-1)
        small = small_dummy
        great_dummy = ListNode(-1)
        great = great_dummy
        
        while head:
            if head.val < x:
                small.next = head
                small = small.next
            else:
                great.next = head
                great = great.next
            head = head.next
            
        # partition two sublists and point the last node to null
        small.next = great_dummy.next
        great.next = None
        
        return small_dummy.next

File: java/PartitionList.java
public class PartitionList {
    /**
     * @param head: The first node of linked list.
     * @param x: an integer
     * @return: a ListNode
     */
    public ListNode partition(ListNode head, int x) {
        if (head == null || head.next == null) return head;

        ListNode smallDummy = new ListNode(0), greatDummy = new ListNode(0);
        ListNode small = smallDummy, great = greatDummy;
        while (head != null) {
            if (head.val < x) {
                small.next = head;
                small = small.next;
            } else {
                great.next = head;
                great = great.next;
            }
            head = head.next;
        }
        small.next = greatDummy.next;
        great.next = null;

        return smallDummy.next;
    }
}
